### PR TITLE
Parameters implementation

### PR DESCRIPTION
**Pull request by Ubermensch01**
NB!: This PR relies on the step_implementation code.

Support for various types of parameters implemented.

### CODE REVIEW

File: example/example_parameters_test.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package example_test
```
[the remark this answers is not in the document]
Comment (Ubermensch01) [reply example/example_parameters_test.go]: *_test is golang's accepted approach to marking test files along with naming convention

File: parameter.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+func convertMapToParameters(parameters map[string]interface{}) []Parameter {
+	result := make([]Parameter, 0)
+
+	for k, v := range parameters {
+		currentParameter := Parameter{
+			Name: k,
+		}
+
+		switch v.(type) {
```
[the remark this answers is not in the document]
Comment (Ubermensch01) [reply parameter.go]: Done.

File: result.go [hunk 3]
```diff
@@ -31,54 +33,82 @@ type Result struct {
...
 
-//Test execute the test and creates an Allure result used by Allure reports
-func Test(t *testing.T, description string, testFunc func()) {
-	wsd = os.Getenv("ALLURE_RESULTS_PATH")
+const (
+	ALLURE_RESULTS_PATH = "ALLURE_RESULTS_PATH"
+	nodeKey             = "current_step_container"
+)
+
+func TestWithParameters(t *testing.T, description string, parameters map[string]interface{}, testFunc func()) {
```
[the remark this answers is not in the document]
Comment (Ubermensch01) [reply result.go]: Done.

File: result.go [hunk 4]
```diff
@@ -31,54 +33,82 @@ type Result struct {
...
 	Stop          int64          `json:"stop,omitempty"`
-	Steps         []Step         `json:"steps,omitempty"`
+	Steps         []stepObject   `json:"steps,omitempty"`
 	Attachments   []Attachment   `json:"attachments,omitempty"`
 }
 
-type Parameter struct {
-	Name  string `json:"name,omitempty"`
-	Value string `json:"value,omitempty"`
+type HasSteps interface {
```
[the remark this answers is not in the document]
Comment (Ubermensch01) [reply result.go]: Done. This interface enables children step objects manipulation. We can have either the test result object or the step object required to possess these functions.

File: example/example_parameters_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaximeBellou) [reply example/example_parameters_test.go]: For files, yes, but not for packages : https://blog.golang.org/package-names
Furthermore, the purpose of the files in this directory is more to provide examples, than to test the code (I mean with automation).

File: example/example_parameters_test.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+		func() {
+			allure.Step(fmt.Sprintf("Number: %d", parameters["int"]), func() {})
+			allure.Step(fmt.Sprintf("String: %s", parameters["string"]), func() {})
+			allure.Step(fmt.Sprintf("Interface: %+v", parameters["struct"]), func() {})
+		})
+}
+
+func TestAllureParameterizedFailing(t *testing.T) {
+	allure.TestWithParameters(t,
+		"This is a test to show allure implementation with a passing test",
```
Comment (MaximeBellou) [inline example/example_parameters_test.go]: Why does the name of the func mention Failing, while the description mentions passing test?

File: result.go [hunk 6]
```diff
@@ -49,7 +48,9 @@ type Before struct {
 	Attachments   []Attachment   `json:"attachments,omitempty"`
 }
 
-type HasSteps interface {
+// This interface provides functions required to manipulate children step records, used in the result object and
+// step object for recursive handling
+type hasSteps interface {
```
Comment (MaximeBellou) [inline result.go]: if you make it private, the comment is not necessary. It is necessary for public methods. Please have a look at https://blog.golang.org/godoc-documenting-go-code

File: step.go [hunk 7]
```diff
@@ -24,15 +24,25 @@ func (s *stepObject) AddStep(step stepObject) {
 	s.ChildrenSteps = append(s.ChildrenSteps, step)
 }
 
+// allure.Step is meant to be wrapped around actions
```
Comment (MaximeBellou) [inline step.go]: The comment must always start by the name of the method. In this case:
// Step is meant .....

File: example/example_parameters_test.go (hunk 5 above)
Comment (Ubermensch01) [reply example/example_parameters_test.go]: This test function doesn't provide any substantial value anyway, removing it.

**Pull request merged**